### ISSUE TITLE
Multi-Thread usage -> Crash due to desynchronized Isolates (pointers desync)

### ISSUE THREAD

**Issue by matiwinnetou** [eclipsesource/J2V8#313]
A recent work of significantly increasing scalability of J2V8 by removing a lock between V8 instances and letting JAVA layer's lock V8 instance and isolate them caused number of issues (crashes). Some issues have been already patched (https://github.com/eclipsesource/J2V8/commit/8f8625015bc1314ee71cfab725d5d72f44979109) and others by passing explicitly Isolate in #308 and #310 but one crash is hard to solve due to inability to easily pass explicitly Isolate by using V8 API. There is also possibility that there maybe other places that could cause similar crash.

An investigation effort was launched to find out why it crashes and the following lines of code have been identified:

In class isolate.cpp
```
String::Value::Value(v8::Local<v8::Value> obj) : str_(NULL), length_(0) {
  if (obj.IsEmpty()) return;
  i::Isolate* isolate = i::Isolate::Current();
  Isolate* v8_isolate = reinterpret_cast<Isolate*>(isolate);
  ENTER_V8(isolate);
  i::HandleScope scope(isolate);
  Local<Context> context = v8_isolate->GetCurrentContext();
  TryCatch try_catch(v8_isolate);
  Local<String> str;
  if (!obj->ToString(context).ToLocal(&str)) return;
  length_ = str->Length();
  str_ = i::NewArray<uint16_t>(length_ + 1);
  str->Write(str_);
}
```

Clearly as with issues #308 and #310 the isolate retrieved via a static method is from time tpo time not the correct one. This suggests a race condition or generally some threading issue. Further investigation shows that Isolates in V8 are in fact stored on thread locals. Even though many places V8's code have been updated to allow passing Isolate explicitly to methods and constructors, as it is visible the constructor that creates String::Value has not been updated or it was not supposed to be updated (unclear). Reading V8 code further reveals places where those ThreadLocals are managed:
isolate.h

```
  // Find the PerThread for this particular (isolate, thread) combination
  // If one does not yet exist, return null.
  PerIsolateThreadData* FindPerThreadDataForThisThread();

  // Find the PerThread for given (isolate, thread) combination
  // If one does not yet exist, return null.
  PerIsolateThreadData* FindPerThreadDataForThread(ThreadId thread_id);

  // Discard the PerThread for this particular (isolate, thread) combination
  // If one does not yet exist, no-op.
  void DiscardPerThreadDataForThisThread();
```
There is also a comment from the authors of V8 that the notion of accessing Isolates via ThreadLocals seems to be deprecated. Sadly, even as of 22 Jul 2017, master version of V8 still contains this TODO and the error prone and brittle accessing of Isolates via ThreadLocals in still the code.

This code seems to prove that the concept of using Isolate::Current() and probably Isolate::GetCurrent() is deprecated:

```
  // TODO(jochen): Once we got rid of Isolate::Current(), we can remove this.
  Isolate::Scope isolate_scope(v8_isolate);
  if (params.entry_hook || !i::Snapshot::Initialize(isolate)) {
    isolate->Init(NULL);
  }
```

Due to Isolate::Current() it also looks that there is hacky code which does not bring confidence that V8 could flawlessly operate in the multi-threaded environment.

Having said that it is not 100% clear if the actual bug is in J2V8 code by misusing V8 API or if the crash happens and a defect is in fact in V8 code that thread locals do not maintain a proper state and hold the correct pointer of Isolate at all times of the lifecycle of the V8.

What does not help in the situation is that V8's locks are binary and V8's are counting semaphores. This is a different paradigm between the two and in theory could cause desynchronization of those locks.

IMHO, there are a few schools of thought how to continue further:

1. Find out why V8 and J2V8 Isolates pointers get desychronized and fix the problem. This is assuming that defect is in the J2V8 code and that J2V8 is incorrectly using V8 API.  One possibility here could be to turn J2V8 locks also in counting semaphores so that models of locks stay in sync (despite the fact that counting semaphores are way complicated than binary ones.

2. Patch V8 in such a way that actually refactoring is performed to remove i::Isolate::Current(); and Isolate::GetCurrent() as they want this anyway, see their TODO entry above. Once refactoring is done, submit PR to V8 codebase. Replace all places where Isolate is retrieved via static method and usage of ThreadLocals with explicitly passing Isolate (tedious code and really verbose but maybe less brittle). Here the bare minimum would be find all places where J2V8 relies on methods that in fact used Isolate::Current, etc and remove them by passing Isolate explicitly. Stretch goal for PR to V8 could be in fact to fix all places where such code is used in the code-base - removing altogether completely Isolate::GetCurrent();

3. Consider some sort of silly patch and workaround that sets the Isolate pointer from J2V8's v8Runtime on ThreadLocal upon every high level entry to V8 method, this would mean that calls to Isolate::Current would be correct. Such hacky patch ideally should also destroy thread local before setting it and clear and or possibly clear that again just before leaving the top level API method call. This is a bit mine field though.

4. Refactor J2V8 code in such a way that it avoids using classes, constructors and methods that take Isolate from ThreadLocal like fire. This may require that one needs to fine alternatives to write code, e.g. turn Local<String> into some other String than String::String::Value (which has a busted constructor that retrieves implicitly Isolate)

Crash report:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fc4c9a6b872, pid=38, tid=0x00007fc4cb117700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_131-b11) (build 1.8.0_131-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libj2v8_linux_x86_64.so+0x56f872]  v8::String::Value::Value(v8::Local<v8::Value>)+0x52
#
# Core dump written. Default location: //core or core.38
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  T H R E A D  ---------------

Current thread (0x00007fc36000b800):  JavaThread "http-nio-8080-exec-20" daemon [_thread_in_native, id=156, stack(0x00007fc4cb017000,0x00007fc4cb118000)]

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000008ce8

Registers:
RAX=0x0000000000000000, RBX=0x0000000000000000, RCX=0x0000000000000000, RDX=0x00007fc4cb117a30
RSP=0x00007fc4cb1154a0, RBP=0x00007fc4cb115540, RSI=0x00007fc3ec05d0a8, RDI=0x0000000000000002
R8 =0x0000000000000000, R9 =0x00000359975ba600, R10=0x00007fc508f55a08, R11=0x00001e1900000000
R12=0x00007fc4cb1154d0, R13=0x00000000f858bb40, R14=0x00007fc4cb115990, R15=0x00007fc3ec05d0a8
RIP=0x00007fc4c9a6b872, EFLAGS=0x0000000000010246, CSGSFS=0x0000000000000033, ERR=0x0000000000000004
  TRAPNO=0x000000000000000e

Top of Stack: (sp=0x00007fc4cb1154a0)
0x00007fc4cb1154a0:   0000000000000005 00007fc4cb115510
0x00007fc4cb1154b0:   00007fc4cb1154e0 00003efbe57060c7
0x00007fc4cb1154c0:   0000000000000002 00003efbe5706001
0x00007fc4cb1154d0:   00007fc4cb1154c0 0000000000000030
0x00007fc4cb1154e0:   000000000000bba1 10f049c554aa7700
0x00007fc4cb1154f0:   00002a8ba937af31 00007fc3ec05d0a8
0x00007fc4cb115500:   0000000000000000 10f049c554aa7700
0x00007fc4cb115510:   00007fc3ec054b40 00000000c53e7db8
0x00007fc4cb115520:   0000000000000000 00000000f858bb40
0x00007fc4cb115530:   00007fc3ec054b40 00007fc36000b800
0x00007fc4cb115540:   00007fc4cb1159c0 00007fc4c9a268e4
0x00007fc4cb115550:   00007fc4cb115598 00000000e570ed5a
0x00007fc4cb115560:   00007fc4cb115a20 00007fc508f40b10
0x00007fc4cb115570:   00007fc4cb1159f0 00007fc36000b9f8
0x00007fc4cb115580:   00002a79af68beb1 0000000c00000000
0x00007fc4cb115590:   0000000c00000000 0000000c00000000
0x00007fc4cb1155a0:   00003efbe5736961 0000000c00000000
0x00007fc4cb1155b0:   00007fc3ec05d0a8 00007fc3ec05d0a8
0x00007fc4cb1155c0:   0000000000000000 00002a8ba937ace9
0x00007fc4cb1155d0:   00007fc508f4cd50 0000000000000003
0x00007fc4cb1155e0:   00007fc4cab4ec6c 00007fc4cb115718
0x00007fc4cb1155f0:   00002a252da3d6f1 0000000100000000
0x00007fc4cb115600:   00007fc4cb1156e0 00007fc4c9c37223
0x00007fc4cb115610:   00007fc4cb115620 00007fc4c9d47379
0x00007fc4cb115620:   00007fc4cb115680 00002a252da04241
0x00007fc4cb115630:   00007fc3ec064720 00000003c9d4703c
0x00007fc4cb115640:   00007fc508f4cd50 00007fc3ec05d0b8
0x00007fc4cb115650:   0000000000000000 0000000000000000
0x00007fc4cb115660:   00007fc3ec05d0a8 00007fc3ec05d0a8
0x00007fc4cb115670:   00000359975ba6c9 00000359975ba6c9
0x00007fc4cb115680:   00000359975ba6c9 00000359975ba6c9
0x00007fc4cb115690:   00000359975ba6c8 000026df57c04ca9 

Instructions: (pc=0x00007fc4c9a6b872)
0x00007fc4c9a6b852:   84 f5 00 00 00 48 8b 05 e2 23 0a 01 49 89 fe 49
0x00007fc4c9a6b862:   89 f7 4c 8d 65 90 8b 38 e8 21 d0 f6 ff 48 89 c3
0x00007fc4c9a6b872:   44 8b a8 e8 8c 00 00 48 89 85 70 ff ff ff c7 80
0x00007fc4c9a6b882:   e8 8c 00 00 03 00 00 00 48 8b 80 28 8d 00 00 48 

Register to memory mapping:

RAX=0x0000000000000000 is an unknown value
RBX=0x0000000000000000 is an unknown value
RCX=0x0000000000000000 is an unknown value
RDX=0x00007fc4cb117a30 is pointing into the stack for thread: 0x00007fc36000b800
RSP=0x00007fc4cb1154a0 is pointing into the stack for thread: 0x00007fc36000b800
RBP=0x00007fc4cb115540 is pointing into the stack for thread: 0x00007fc36000b800
RSI=0x00007fc3ec05d0a8 is an unknown value
RDI=0x0000000000000002 is an unknown value
R8 =0x0000000000000000 is an unknown value
R9 =0x00000359975ba600 is an unknown value
R10=0x00007fc508f55a08 is an unknown value
R11=0x00001e1900000000 is an unknown value
R12=0x00007fc4cb1154d0 is pointing into the stack for thread: 0x00007fc36000b800
R13=0x00000000f858bb40 is an oop
[Ljava.lang.Object; 
 - klass: 'java/lang/Object'[]
 - length: 3
R14=0x00007fc4cb115990 is pointing into the stack for thread: 0x00007fc36000b800
R15=0x00007fc3ec05d0a8 is an unknown value


Stack: [0x00007fc4cb017000,0x00007fc4cb118000],  sp=0x00007fc4cb1154a0,  free space=1017k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libj2v8_linux_x86_64.so+0x56f872]  v8::String::Value::Value(v8::Local<v8::Value>)+0x52
C  [libj2v8_linux_x86_64.so+0x52a8e4]  getResult(JNIEnv_*, _jobject*&, long, v8::Local<v8::Value>&, int)+0x732
C  [libj2v8_linux_x86_64.so+0x518be4]  Java_com_eclipsesource_v8_V8__1executeFunction__JIJLjava_lang_String_2J+0x1fb
J 10189  com.eclipsesource.v8.V8._executeFunction(JIJLjava/lang/String;J)Ljava/lang/Object; (0 bytes) @ 0x00007fc4f8b9b14d [0x00007fc4f8b9b080+0xcd]
J 12892 C2 com.eclipsesource.v8.V8Object.executeJSFunction(Ljava/lang/String;[Ljava/lang/Object;)Ljava/lang/Object; (279 bytes) @ 0x00007fc4f9fff508 [0x00007fc4f9fff1c0+0x348]
J 13578 C2 pubse.ecs.system.renderer.servlet.j2v8.PoolingJ2V8JavaScriptRenderer.apply(Lde/mobile/ecs/World;)V (340 bytes) @ 0x00007fc4fab1cfa4 [0x00007fc4fab1c780+0x824]
J 13092 C2 pubse.ecs.system.renderer.ResponseRenderer.apply(Lde/mobile/ecs/World;)V (79 bytes) @ 0x00007fc4fa92d3b4 [0x00007fc4fa92d020+0x394]
J 11516 C2 com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed()Ljava/lang/Object; (91 bytes) @ 0x00007fc4fa680350 [0x00007fc4fa67a3e0+0x5f70]
J 13634 C2 pubse.config.InboundModule$Monitor.invoke(Lorg/aopalliance/intercept/MethodInvocation;)Ljava/lang/Object; (70 bytes) @ 0x00007fc4f9cea5bc [0x00007fc4f9cea4a0+0x11c]
J 11516 C2 com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed()Ljava/lang/Object; (91 bytes) @ 0x00007fc4fa67a560 [0x00007fc4fa67a3e0+0x180]
J 12832 C2 pubse.tomcat.handler.ServletRouter$$Lambda$188.handle(Ljavax/servlet/http/HttpServletRequest;Ljavax/servlet/http/HttpServletResponse;)V (10 bytes) @ 0x00007fc4faa54520 [0x00007fc4faa53be0+0x940]
J 11093 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fc4f969e020 [0x00007fc4f969dee0+0x140]
J 12551 C2 org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (402 bytes) @ 0x00007fc4fa992a80 [0x00007fc4fa9917c0+0x12c0]
J 11093 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fc4f969e328 [0x00007fc4f969dee0+0x448]
J 12771 C2 pubse.util.ServletFilters$Utf8EncodingFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (21 bytes) @ 0x00007fc4faa25180 [0x00007fc4faa24da0+0x3e0]
J 11093 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fc4f969e328 [0x00007fc4f969dee0+0x448]
J 12707 C2 pubse.util.ServletFilters$LoggingFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (76 bytes) @ 0x00007fc4fa3bcfe8 [0x00007fc4fa3bcbe0+0x408]
J 11093 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fc4f969e328 [0x00007fc4f969dee0+0x448]
J 13553 C2 org.apache.catalina.core.StandardWrapperValve.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (1389 bytes) @ 0x00007fc4fa464fec [0x00007fc4fa464e00+0x1ec]
J 12726 C2 org.apache.catalina.authenticator.AuthenticatorBase.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (903 bytes) @ 0x00007fc4fa5b2900 [0x00007fc4fa5b1980+0xf80]
J 12739 C2 org.apache.catalina.core.StandardHostValve.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (402 bytes) @ 0x00007fc4faa03420 [0x00007fc4faa030c0+0x360]
J 13494 C2 org.apache.catalina.connector.CoyoteAdapter.service(Lorg/apache/coyote/Request;Lorg/apache/coyote/Response;)V (805 bytes) @ 0x00007fc4fa105994 [0x00007fc4fa1054a0+0x4f4]
J 13979 C2 org.apache.coyote.http11.AbstractHttp11Processor.process(Lorg/apache/tomcat/util/net/SocketWrapper;)Lorg/apache/tomcat/util/net/AbstractEndpoint$Handler$SocketState; (1128 bytes) @ 0x00007fc4f9b0f65c [0x00007fc4f9b0f020+0x63c]
J 13493 C2 org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(Lorg/apache/tomcat/util/net/SocketWrapper;Lorg/apache/tomcat/util/net/SocketStatus;)Lorg/apache/tomcat/util/net/AbstractEndpoint$Handler$SocketState; (1073 bytes) @ 0x00007fc4fa475960 [0x00007fc4fa475480+0x4e0]
J 13115 C2 org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun()V (603 bytes) @ 0x00007fc4fa0607fc [0x00007fc4fa060680+0x17c]
J 12608 C2 org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run()V (79 bytes) @ 0x00007fc4fa49626c [0x00007fc4fa4961c0+0xac]
J 13557% C2 java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V (225 bytes) @ 0x00007fc4fa57c134 [0x00007fc4fa57bf20+0x214]
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5
j  org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run()V+4
j  java.lang.Thread.run()V+11
v  ~StubRoutines::call_stub
V  [libjvm.so+0x691d16]  JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*)+0x1056
V  [libjvm.so+0x692221]  JavaCalls::call_virtual(JavaValue*, KlassHandle, Symbol*, Symbol*, JavaCallArguments*, Thread*)+0x321
V  [libjvm.so+0x6926c7]  JavaCalls::call_virtual(JavaValue*, Handle, KlassHandle, Symbol*, Symbol*, Thread*)+0x47
V  [libjvm.so+0x72da50]  thread_entry(JavaThread*, Thread*)+0xa0
V  [libjvm.so+0xa76833]  JavaThread::thread_main_inner()+0x103
V  [libjvm.so+0xa7697c]  JavaThread::run()+0x11c
V  [libjvm.so+0x927568]  java_start(Thread*)+0x108
```

**Comment by irbull** [comment]
Looking at number 3 in your list, we wouldn't need to set this on ever call to V8. Since the locker is tied to a Runtime (Isolate), whenever we acquire the lock, we could set the isolate on the ThreadLocal. The lock is only valid on a single (Java) thread until it's released.

However, we saw some crashes on Isolate Creation, which would only have ever happened on a single thread, so I'm not sure if this solution would be enough.

**Comment by irbull** [comment]
And regarding number 1 (the locking), binary semaphores are a subset of counting ones (you have the lock if the count is >= 1). However, I have made the assumption that you only acquire / release the lock in Java, so I don't actually check the counting value. In the `checkLock` we could check the native lock to see if we still actually hold it.

Having said that, if we don't actually hold it (i.e. it was released somehow by native code), this would likely be pretty bad. Not sure how we could get into this situation, and it means that simply calling a V8 method would suddenly throw an `InvalidThreadAccess` exception. Also, if we could acquire the lock natively, then we wouldn't even have a Java object to test this on.

**Comment by irbull** [comment]
Sorry, regarding the locker comment, I was wrong. The logic I use to implement the binary semaphore using the counting one is to ignore all calls to `acquire` if the lock is already acquired (count >= 1), and release on the first call to release. So subsequent calls to `acquire` are simply a no-op. 

This has the same affect though, it assumes that the lock can only be twiddled using Java code. This is something we should test.

Having said all this, there were two reports of crashes on creation before we implemented the new locking strategy; but maybe we shouldn't put too much weight on those reports (they could be unrelated).

**Comment by irbull** [comment]
Also, (and maybe you mentioned this is another issue) there are methods on the Isolate called `Enter()` and `Exit()`. We don't use these in J2V8, and maybe we should. When we acquire a lock for an isolate, maybe we should enter the isolate. The implementation of Enter does appear to setup some Thread specific data. I've tried this, and all the tests pass. We could try this, and see if it helps.

**Comment by irbull** [comment]
Here is an example of what I mean: https://github.com/eclipsesource/J2V8/commit/5f6e1a81f56d012b6107cdf339e2441cfaf96f72

**Comment by irbull** [comment]
@matiwinnetou Do you think you could pull this change-set down (it's the tip of this branch https://github.com/eclipsesource/J2V8/tree/isolate_enter_branch) and give it a try? I'm curious if the `Enter` / `Exit` have any effect?

Although, looking closer, maybe not :(. It seems that `Enter` is called when the `Scope` is instantiated, and this happens on every J2V8 call.

**Comment by irbull** [comment]
And there also appears to be a method called `DiscardThreadSpecificMetadata` which says should be used if a thread is terminating and the isolate will outlive the thread. I don't know what that means if an isolate will be moved to a different thread, but maybe we should Discard the metadata when releasing the lock. Again, doing this results in all tests passing.

**Comment by matiwinnetou** [comment]
I can definitely test something once you think it is ready. I saw you pushed 3 commits in an isolate_branch. I put some comments there.

I have tests cases to reproduce this error, it takes about 10 hours for one of our instances in PROD under moderate traffic to crash. Interestingly enough in our test environment where we still run WMWare virtual machine and processors it does not crash. This is really a race condition that occurs in specific circumstances. Such testing in PROD naturally is bit risky so I will be trying to work with DevOps (SiteOps) to isolate (no pun intended) those tests outside of user traffic facing machines.

**Comment by matiwinnetou** [comment]
Reproducing a crash is much much easier (takes about 1 hour in PROD), when I revert back my PRs and in many places code relies on Isolate::Current(). Such a funky build could be used to quickly verify if threading synchronisation and locks works correctly at much faster rate.

**Comment by matiwinnetou** [comment]
I am running tests locally on 4.8.0 release just to see if I can reproduce this Isolate bugs on my local dev box, MacBook Pro, 16 GB RAM (OSX, Sierra). So far rans without any crashes.

I feel obliged to mention that our production machine has 30 CPU cores, the reason I mention this is that perhaps this defect could be is fact in V8. Most V8 setups not only do not so many isolates and threads and and if they do they do not use so many cores.

DevOps work pal is trying to get me a 16 core box in integration system to be able to reliably reproduce crashes without deploying to PROD.

**Comment by matiwinnetou** [comment]
I got a buffy many cores machine, 32 in qa system, testes running on vanilla 4.8.0 to see if I can even reproduce isolate crashes

**Comment by matiwinnetou** [comment]
@irbull

Sadly issue is not fixed yet.

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fd80dbf38d2, pid=38, tid=0x00007fd7ffafc700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_131-b11) (build 1.8.0_131-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libj2v8_linux_x86_64.so+0x56f8d2]  v8::String::Value::Value(v8::Local<v8::Value>)+0x52
#
# Core dump written. Default location: /mnt/mesos/sandbox/core or core.38
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  T H R E A D  ---------------

Current thread (0x00007fd658001800):  JavaThread "http-nio-8080-exec-5" daemon [_thread_in_native, id=142, stack(0x00007fd7ff9fc000,0x00007fd7ffafd000)]

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000008ce8

Registers:
RAX=0x0000000000000000, RBX=0x0000000000000000, RCX=0x0000000000000000, RDX=0x00007fd7ffafca30
RSP=0x00007fd7ffaf9410, RBP=0x00007fd7ffaf94b0, RSI=0x00007fd678079d68, RDI=0x0000000000000002
R8 =0x0000000000000000, R9 =0x00000000ed1e68c0, R10=0x00007fd8528d5978, R11=0x0000000000000000
R12=0x00007fd7ffaf9440, R13=0x00007fd7ffaf9838, R14=0x00007fd7ffaf9730, R15=0x00007fd678079d68
RIP=0x00007fd80dbf38d2, EFLAGS=0x0000000000010246, CSGSFS=0x0000000000000033, ERR=0x0000000000000004
  TRAPNO=0x000000000000000e

Top of Stack: (sp=0x00007fd7ffaf9410)
0x00007fd7ffaf9410:   00007fd700000060 00007fd73c009d10
0x00007fd7ffaf9420:   0000000000000000 0000000000000000
0x00007fd7ffaf9430:   00007fd678079d58 00007fd678079d58
0x00007fd7ffaf9440:   00007fd678079d58 0000000000000000
0x00007fd7ffaf9450:   00007fd7ffaf9480 602378c27ea94300
0x00007fd7ffaf9460:   00007fd7ffaf9490 00007fd678079d68
0x00007fd7ffaf9470:   0000000000000000 602378c27ea94300
0x00007fd7ffaf9480:   00007fd658001800 00000000c7dcf2b0
0x00007fd7ffaf9490:   0000000000000000 00007fd7ffaf9838
0x00007fd7ffaf94a0:   00007fd658001800 00007fd658001800
0x00007fd7ffaf94b0:   00007fd7ffaf9770 00007fd80dba6623
0x00007fd7ffaf94c0:   00007fd68001ae48 00000000000000d8
0x00007fd7ffaf94d0:   00007fd678000f50 00007fd73c001a60
0x00007fd7ffaf94e0:   00007fd7ffaf9780 00007fd6580019f8
0x00007fd7ffaf94f0:   0000000c00000000 00000000000000d8
0x00007fd7ffaf9500:   0000000c00000000 00007fd658001800
0x00007fd7ffaf9510:   00007fd80f2c36e8 00007fd73c009d10
0x00007fd7ffaf9520:   00007fd678079d50 00007fd678079d50
0x00007fd7ffaf9530:   00007fd678079d58 00007fd678079d68
0x00007fd7ffaf9540:   00007fd678079d68 00007fd678079d50
0x00007fd7ffaf9550:   00007fd678079d68 00007fd73c009d10
0x00007fd7ffaf9560:   00007fd73c001a60 00007fd73c009d10
0x00007fd7ffaf9570:   00007fd73c001a70 00007fd73c009d10
0x00007fd7ffaf9580:   00007fd73c063f90 00007fd73c063f90
0x00007fd7ffaf9590:   00007fd73c063f90 00007fd678079d50
0x00007fd7ffaf95a0:   00007fd7ffaf9528 00007fd73c009d10
0x00007fd7ffaf95b0:   00007fd678000f50 00007fd73c009d10
0x00007fd7ffaf95c0:   00007fd73c064070 00007fd73c064070
0x00007fd7ffaf95d0:   00007fd73c064070 00007fd678079d58
0x00007fd7ffaf95e0:   00007fd678079d68 00007fd678079d68
0x00007fd7ffaf95f0:   00003cbb52bab1d9 00003cbb52bab1d9
0x00007fd7ffaf9600:   00003cbb52bab1d9 00003cbb52bab1d9 

Instructions: (pc=0x00007fd80dbf38d2)
0x00007fd80dbf38b2:   84 f5 00 00 00 48 8b 05 82 23 0a 01 49 89 fe 49
0x00007fd80dbf38c2:   89 f7 4c 8d 65 90 8b 38 e8 f1 cf f6 ff 48 89 c3
0x00007fd80dbf38d2:   44 8b a8 e8 8c 00 00 48 89 85 70 ff ff ff c7 80
0x00007fd80dbf38e2:   e8 8c 00 00 03 00 00 00 48 8b 80 28 8d 00 00 48 

Register to memory mapping:

RAX=0x0000000000000000 is an unknown value
RBX=0x0000000000000000 is an unknown value
RCX=0x0000000000000000 is an unknown value
RDX=0x00007fd7ffafca30 is pointing into the stack for thread: 0x00007fd658001800
RSP=0x00007fd7ffaf9410 is pointing into the stack for thread: 0x00007fd658001800
RBP=0x00007fd7ffaf94b0 is pointing into the stack for thread: 0x00007fd658001800
RSI=0x00007fd678079d68 is an unknown value
RDI=0x0000000000000002 is an unknown value
R8 =0x0000000000000000 is an unknown value
R9 =0x00000000ed1e68c0 is an oop
com.eclipsesource.v8.V8Array 
 - klass: 'com/eclipsesource/v8/V8Array'
R10=0x00007fd8528d5978 is at entry_point+56 in (nmethod*)0x00007fd8528d57d0
R11=0x0000000000000000 is an unknown value
R12=0x00007fd7ffaf9440 is pointing into the stack for thread: 0x00007fd658001800
R13=0x00007fd7ffaf9838 is pointing into the stack for thread: 0x00007fd658001800
R14=0x00007fd7ffaf9730 is pointing into the stack for thread: 0x00007fd658001800
R15=0x00007fd678079d68 is an unknown value


Stack: [0x00007fd7ff9fc000,0x00007fd7ffafd000],  sp=0x00007fd7ffaf9410,  free space=1013k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libj2v8_linux_x86_64.so+0x56f8d2]  v8::String::Value::Value(v8::Local<v8::Value>)+0x52
C  [libj2v8_linux_x86_64.so+0x522623]  Java_com_eclipsesource_v8_V8__1arrayGetString+0x705
J 8523  com.eclipsesource.v8.V8._arrayGetString(JJI)Ljava/lang/String; (0 bytes) @ 0x00007fd8528d59ee [0x00007fd8528d5940+0xae]
J 11630 C2 com.eclipsesource.v8.V8.callObjectJavaMethod(JLcom/eclipsesource/v8/V8Object;Lcom/eclipsesource/v8/V8Array;)Ljava/lang/Object; (140 bytes) @ 0x00007fd852825bb0 [0x00007fd852825960+0x250]
v  ~StubRoutines::call_stub
V  [libjvm.so+0x691d16]  JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*)+0x1056
V  [libjvm.so+0x6d3c1b]  jni_invoke_nonstatic(JNIEnv_*, JavaValue*, _jobject*, JNICallType, _jmethodID*, JNI_ArgumentPusher*, Thread*)+0x40b
V  [libjvm.so+0x6d9dd1]  jni_CallObjectMethodV+0x111
C  [libj2v8_linux_x86_64.so+0x52b6ef]  JNIEnv_::CallObjectMethod(_jobject*, _jmethodID*, ...)+0xc7
C  [libj2v8_linux_x86_64.so+0x5261d0]  objectCallback(v8::FunctionCallbackInfo<v8::Value> const&)+0x213
C  [libj2v8_linux_x86_64.so+0xa89411]  v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&))+0x131
C  [libj2v8_linux_x86_64.so+0xadb8bf]  v8::internal::MaybeHandle<v8::internal::Object> v8::internal::(anonymous namespace)::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::HeapObject>, v8::internal::Handle<v8::internal::FunctionTemplateInfo>, v8::internal::Handle<v8::internal::Object>, v8::internal::BuiltinArguments)+0x25f
C  [libj2v8_linux_x86_64.so+0xadc682]  v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*)+0xf2
C  0x000019f8a5e060c7
C  0x000019f8a616faaf
C  0x000019f8a5e0ed7d
C  0x000019f8a5e36ba1
C  [libj2v8_linux_x86_64.so+0x73b283]  v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*)+0x1d3
C  [libj2v8_linux_x86_64.so+0x554be2]  v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*)+0x1c2
C  [libj2v8_linux_x86_64.so+0x56c641]  v8::Function::Call(v8::Local<v8::Value>, int, v8::Local<v8::Value>*)+0x41
C  [libj2v8_linux_x86_64.so+0x5186fe]  invokeFunction(JNIEnv_*, v8::Isolate*, long&, long&, _jstring*&, long&, v8::Local<v8::Value>&)+0x376
C  [libj2v8_linux_x86_64.so+0x518c08]  Java_com_eclipsesource_v8_V8__1executeFunction__JIJLjava_lang_String_2J+0x1bc
J 10439  com.eclipsesource.v8.V8._executeFunction(JIJLjava/lang/String;J)Ljava/lang/Object; (0 bytes) @ 0x00007fd85252780d [0x00007fd852527740+0xcd]
J 12603 C2 com.eclipsesource.v8.V8Object.executeJSFunction(Ljava/lang/String;[Ljava/lang/Object;)Ljava/lang/Object; (279 bytes) @ 0x00007fd85341a2c8 [0x00007fd853419f80+0x348]
J 12801 C2 pubse.ecs.system.renderer.servlet.j2v8.PoolingJ2V8JavaScriptRenderer.apply(Lde/mobile/ecs/World;)V (340 bytes) @ 0x00007fd8531e9fb0 [0x00007fd8531e96a0+0x910]
J 12894 C2 pubse.ecs.system.renderer.ResponseRenderer.apply(Lde/mobile/ecs/World;)V (79 bytes) @ 0x00007fd851d405f4 [0x00007fd851d40260+0x394]
J 12021 C2 com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed()Ljava/lang/Object; (91 bytes) @ 0x00007fd85329952c [0x00007fd853293980+0x5bac]
J 12558 C2 pubse.config.InboundModule$Monitor.invoke(Lorg/aopalliance/intercept/MethodInvocation;)Ljava/lang/Object; (70 bytes) @ 0x00007fd8534004dc [0x00007fd8534003c0+0x11c]
J 12021 C2 com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed()Ljava/lang/Object; (91 bytes) @ 0x00007fd853293b28 [0x00007fd853293980+0x1a8]
J 12601 C2 pubse.tomcat.handler.ServletRouter$$Lambda$179.handle(Ljavax/servlet/http/HttpServletRequest;Ljavax/servlet/http/HttpServletResponse;)V (10 bytes) @ 0x00007fd853416f34 [0x00007fd853416700+0x834]
J 11026 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fd8530680f4 [0x00007fd853067ca0+0x454]
J 11569 C2 org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (402 bytes) @ 0x00007fd8531a8b20 [0x00007fd8531a7880+0x12a0]
J 11026 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fd853067fb0 [0x00007fd853067ca0+0x310]
J 12147 C2 pubse.util.ServletFilters$Utf8EncodingFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (21 bytes) @ 0x00007fd85332fe80 [0x00007fd85332faa0+0x3e0]
J 11026 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fd853067fb0 [0x00007fd853067ca0+0x310]
J 12266 C2 pubse.util.ServletFilters$LoggingFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (76 bytes) @ 0x00007fd852c0dde8 [0x00007fd852c0da80+0x368]
J 11026 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fd853067fb0 [0x00007fd853067ca0+0x310]
J 12970 C2 org.apache.catalina.core.StandardWrapperValve.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (1389 bytes) @ 0x00007fd853540cec [0x00007fd853540b00+0x1ec]
J 12062 C2 org.apache.catalina.authenticator.AuthenticatorBase.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (903 bytes) @ 0x00007fd852995768 [0x00007fd852994b00+0xc68]
J 12074 C2 org.apache.catalina.core.StandardHostValve.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (402 bytes) @ 0x00007fd8532c9760 [0x00007fd8532c9400+0x360]
J 12539 C2 org.apache.catalina.connector.CoyoteAdapter.service(Lorg/apache/coyote/Request;Lorg/apache/coyote/Response;)V (805 bytes) @ 0x00007fd8533f8b14 [0x00007fd8533f8600+0x514]
J 12928 C2 org.apache.coyote.http11.AbstractHttp11Processor.process(Lorg/apache/tomcat/util/net/SocketWrapper;)Lorg/apache/tomcat/util/net/AbstractEndpoint$Handler$SocketState; (1128 bytes) @ 0x00007fd85350f0a8 [0x00007fd85350ee40+0x268]
J 12531 C2 org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(Lorg/apache/tomcat/util/net/SocketWrapper;Lorg/apache/tomcat/util/net/SocketStatus;)Lorg/apache/tomcat/util/net/AbstractEndpoint$Handler$SocketState; (1073 bytes) @ 0x00007fd8533d6738 [0x00007fd8533d6280+0x4b8]
J 11747 C2 org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun()V (603 bytes) @ 0x00007fd853202950 [0x00007fd8532027e0+0x170]
J 11717 C2 org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run()V (79 bytes) @ 0x00007fd8519e0fac [0x00007fd8519e0f00+0xac]
j  java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+95
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5
j  org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run()V+4
j  java.lang.Thread.run()V+11
v  ~StubRoutines::call_stub
V  [libjvm.so+0x691d16]  JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*)+0x1056
V  [libjvm.so+0x692221]  JavaCalls::call_virtual(JavaValue*, KlassHandle, Symbol*, Symbol*, JavaCallArguments*, Thread*)+0x321
V  [libjvm.so+0x6926c7]  JavaCalls::call_virtual(JavaValue*, Handle, KlassHandle, Symbol*, Symbol*, Thread*)+0x47
V  [libjvm.so+0x72da50]  thread_entry(JavaThread*, Thread*)+0xa0
V  [libjvm.so+0xa76833]  JavaThread::thread_main_inner()+0x103
V  [libjvm.so+0xa7697c]  JavaThread::run()+0x11c
V  [libjvm.so+0x927568]  java_start(Thread*)+0x108
C  [libpthread.so.0+0x76ba]  start_thread+0xca

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
J 8523  com.eclipsesource.v8.V8._arrayGetString(JJI)Ljava/lang/String; (0 bytes) @ 0x00007fd8528d5978 [0x00007fd8528d5940+0x38]
J 11630 C2 com.eclipsesource.v8.V8.callObjectJavaMethod(JLcom/eclipsesource/v8/V8Object;Lcom/eclipsesource/v8/V8Array;)Ljava/lang/Object; (140 bytes) @ 0x00007fd852825bb0 [0x00007fd852825960+0x250]
v  ~StubRoutines::call_stub
J 10439  com.eclipsesource.v8.V8._executeFunction(JIJLjava/lang/String;J)Ljava/lang/Object; (0 bytes) @ 0x00007fd852527793 [0x00007fd852527740+0x53]
J 12603 C2 com.eclipsesource.v8.V8Object.executeJSFunction(Ljava/lang/String;[Ljava/lang/Object;)Ljava/lang/Object; (279 bytes) @ 0x00007fd85341a2c8 [0x00007fd853419f80+0x348]
J 12801 C2 pubse.ecs.system.renderer.servlet.j2v8.PoolingJ2V8JavaScriptRenderer.apply(Lde/mobile/ecs/World;)V (340 bytes) @ 0x00007fd8531e9fb0 [0x00007fd8531e96a0+0x910]
J 12894 C2 pubse.ecs.system.renderer.ResponseRenderer.apply(Lde/mobile/ecs/World;)V (79 bytes) @ 0x00007fd851d405f4 [0x00007fd851d40260+0x394]
J 12021 C2 com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed()Ljava/lang/Object; (91 bytes) @ 0x00007fd85329952c [0x00007fd853293980+0x5bac]
J 12558 C2 pubse.config.InboundModule$Monitor.invoke(Lorg/aopalliance/intercept/MethodInvocation;)Ljava/lang/Object; (70 bytes) @ 0x00007fd8534004dc [0x00007fd8534003c0+0x11c]
J 12021 C2 com.google.inject.internal.InterceptorStackCallback$InterceptedMethodInvocation.proceed()Ljava/lang/Object; (91 bytes) @ 0x00007fd853293b28 [0x00007fd853293980+0x1a8]
J 12601 C2 pubse.tomcat.handler.ServletRouter$$Lambda$179.handle(Ljavax/servlet/http/HttpServletRequest;Ljavax/servlet/http/HttpServletResponse;)V (10 bytes) @ 0x00007fd853416f34 [0x00007fd853416700+0x834]
J 11026 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fd8530680f4 [0x00007fd853067ca0+0x454]
J 11569 C2 org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (402 bytes) @ 0x00007fd8531a8b20 [0x00007fd8531a7880+0x12a0]
J 11026 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fd853067fb0 [0x00007fd853067ca0+0x310]
J 12147 C2 pubse.util.ServletFilters$Utf8EncodingFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (21 bytes) @ 0x00007fd85332fe80 [0x00007fd85332faa0+0x3e0]
J 11026 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fd853067fb0 [0x00007fd853067ca0+0x310]
J 12266 C2 pubse.util.ServletFilters$LoggingFilter.doFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;Ljavax/servlet/FilterChain;)V (76 bytes) @ 0x00007fd852c0dde8 [0x00007fd852c0da80+0x368]
J 11026 C2 org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(Ljavax/servlet/ServletRequest;Ljavax/servlet/ServletResponse;)V (574 bytes) @ 0x00007fd853067fb0 [0x00007fd853067ca0+0x310]
J 12970 C2 org.apache.catalina.core.StandardWrapperValve.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (1389 bytes) @ 0x00007fd853540cec [0x00007fd853540b00+0x1ec]
J 12062 C2 org.apache.catalina.authenticator.AuthenticatorBase.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (903 bytes) @ 0x00007fd852995768 [0x00007fd852994b00+0xc68]
J 12074 C2 org.apache.catalina.core.StandardHostValve.invoke(Lorg/apache/catalina/connector/Request;Lorg/apache/catalina/connector/Response;)V (402 bytes) @ 0x00007fd8532c9760 [0x00007fd8532c9400+0x360]
J 12539 C2 org.apache.catalina.connector.CoyoteAdapter.service(Lorg/apache/coyote/Request;Lorg/apache/coyote/Response;)V (805 bytes) @ 0x00007fd8533f8b14 [0x00007fd8533f8600+0x514]
J 12928 C2 org.apache.coyote.http11.AbstractHttp11Processor.process(Lorg/apache/tomcat/util/net/SocketWrapper;)Lorg/apache/tomcat/util/net/AbstractEndpoint$Handler$SocketState; (1128 bytes) @ 0x00007fd85350f0a8 [0x00007fd85350ee40+0x268]
J 12531 C2 org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(Lorg/apache/tomcat/util/net/SocketWrapper;Lorg/apache/tomcat/util/net/SocketStatus;)Lorg/apache/tomcat/util/net/AbstractEndpoint$Handler$SocketState; (1073 bytes) @ 0x00007fd8533d6738 [0x00007fd8533d6280+0x4b8]
J 11747 C2 org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun()V (603 bytes) @ 0x00007fd853202950 [0x00007fd8532027e0+0x170]
J 11717 C2 org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run()V (79 bytes) @ 0x00007fd8519e0fac [0x00007fd8519e0f00+0xac]
j  java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+95
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5
j  org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run()V+4
j  java.lang.Thread.run()V+11
v  ~StubRoutines::call_stub

---------------  P R O C E S S  ---------------

Java Threads: ( => current thread )
  0x00007fd6a8019800 JavaThread "Retrofit-Idle" [_thread_blocked, id=284, stack(0x00007fd7ffeff000,0x00007fd800000000)]
  0x00007fd6b400e800 JavaThread "Retrofit-Idle" [_thread_blocked, id=283, stack(0x00007fd7fe4f1000,0x00007fd7fe5f2000)]
  0x00007fd6b8014800 JavaThread "Retrofit-Idle" [_thread_blocked, id=282, stack(0x00007fd80ef9d000,0x00007fd80f09e000)]
  0x00007fd6ac00f800 JavaThread "Retrofit-Idle" [_thread_blocked, id=281, stack(0x00007fd7fe3f0000,0x00007fd7fe4f1000)]
  0x00007fd6a0009000 JavaThread "Retrofit-Idle" [_thread_blocked, id=280, stack(0x00007fd7fe2ef000,0x00007fd7fe3f0000)]
  0x00007fd6a400b800 JavaThread "Retrofit-Idle" daemon [_thread_blocked, id=277, stack(0x00007fd8045f0000,0x00007fd8046f1000)]
  0x00007fd66c24a000 JavaThread "RxIoScheduler-20" daemon [_thread_blocked, id=275, stack(0x00007fd8048f3000,0x00007fd8049f4000)]
  0x00007fd658004800 JavaThread "http-nio-8080-exec-10" daemon [_thread_blocked, id=244, stack(0x00007fd80f19f000,0x00007fd80f2a0000)]
  0x00007fd664005800 JavaThread "http-nio-8080-exec-11" daemon [_thread_blocked, id=243, stack(0x00007fd8046f1000,0x00007fd8047f2000)]
  0x00007fd658003800 JavaThread "http-nio-8080-exec-9" daemon [_thread_blocked, id=187, stack(0x00007fd80f5a1000,0x00007fd80f6a2000)]
  0x00007fd664004800 JavaThread "http-nio-8080-exec-8" daemon [_thread_blocked, id=186, stack(0x00007fd7ffbfe000,0x00007fd7ffcff000)]
  0x00007fd664002000 JavaThread "http-nio-8080-exec-7" daemon [_thread_blocked, id=144, stack(0x00007fd80f09e000,0x00007fd80f19f000)]
  0x00007fd664001000 JavaThread "http-nio-8080-exec-6" daemon [_thread_blocked, id=143, stack(0x00007fd7ff8fb000,0x00007fd7ff9fc000)]
=>0x00007fd658001800 JavaThread "http-nio-8080-exec-5" daemon [_thread_in_native, id=142, stack(0x00007fd7ff9fc000,0x00007fd7ffafd000)]
  0x00007fd86000a000 JavaThread "DestroyJavaVM" [_thread_blocked, id=39, stack(0x00007fd867a4e000,0x00007fd867b4f000)]
  0x00007fd680098000 JavaThread "http-nio-8080-Acceptor-0" daemon [_thread_in_native, id=139, stack(0x00007fd7feef3000,0x00007fd7feff4000)]
  0x00007fd680096000 JavaThread "http-nio-8080-ClientPoller-1" daemon [_thread_in_native, id=138, stack(0x00007fd7feff4000,0x00007fd7ff0f5000)]
  0x00007fd68007b800 JavaThread "http-nio-8080-ClientPoller-0" daemon [_thread_in_native, id=137, stack(0x00007fd7ff0f5000,0x00007fd7ff1f6000)]
  0x00007fd68005b000 JavaThread "http-nio-8080-exec-4" daemon [_thread_blocked, id=136, stack(0x00007fd7ff1f6000,0x00007fd7ff2f7000)]
  0x00007fd680059800 JavaThread "http-nio-8080-exec-3" daemon [_thread_blocked, id=135, stack(0x00007fd7ff2f7000,0x00007fd7ff3f8000)]
  0x00007fd680058000 JavaThread "http-nio-8080-exec-2" daemon [_thread_blocked, id=134, stack(0x00007fd7ff3f8000,0x00007fd7ff4f9000)]
  0x00007fd680056800 JavaThread "http-nio-8080-exec-1" daemon [_thread_blocked, id=133, stack(0x00007fd7ff4f9000,0x00007fd7ff5fa000)]
  0x00007fd680052800 JavaThread "ContainerBackgroundProcessor[StandardEngine[Tomcat]]" daemon [_thread_blocked, id=132, stack(0x00007fd7ff5fa000,0x00007fd7ff6fb000)]
  0x00007fd68004b800 JavaThread "NioBlockingSelector.BlockPoller-1" daemon [_thread_in_native, id=129, stack(0x00007fd7ffafd000,0x00007fd7ffbfe000)]
  0x00007fd6b8010800 JavaThread "RxIoScheduler-1 (Evictor)" daemon [_thread_blocked, id=115, stack(0x00007fd8049f4000,0x00007fd804af5000)]
  0x00007fd860b14800 JavaThread "IzmoPictureResolver" [_thread_blocked, id=114, stack(0x00007fd804af5000,0x00007fd804bf6000)]
  0x00007fd860b13000 JavaThread "MakeModelAlternativesResolver" [_thread_blocked, id=113, stack(0x00007fd804bf6000,0x00007fd804cf7000)]
  0x00007fd860b12000 JavaThread "RemoteRefDataProvider" [_thread_blocked, id=112, stack(0x00007fd804cf7000,0x00007fd804df8000)]
  0x00007fd860ef7000 JavaThread "MakeModelResolver" [_thread_blocked, id=111, stack(0x00007fd804df8000,0x00007fd804ef9000)]
  0x00007fd860ef6000 JavaThread "GeoLocationResolver" [_thread_blocked, id=110, stack(0x00007fd804ef9000,0x00007fd804ffa000)]
  0x00007fd860e78800 JavaThread "AbTestResolver" [_thread_blocked, id=109, stack(0x00007fd80c1ff000,0x00007fd80c300000)]
  0x00007fd6b8001800 JavaThread "metrics-graphite-reporter-1-thread-1" daemon [_thread_blocked, id=108, stack(0x00007fd80c0fe000,0x00007fd80c1ff000)]
  0x00007fd860b6f000 JavaThread "commons-pool-EvictionTimer" daemon [_thread_blocked, id=92, stack(0x00007fd80f4a0000,0x00007fd80f5a1000)]
  0x00007fd860e25800 JavaThread "Connection evictor" daemon [_thread_blocked, id=90, stack(0x00007fd80faa2000,0x00007fd80fba3000)]
  0x00007fd8605d2800 JavaThread "Log4j2-TF-1-AsyncLogger[AsyncContext@42a57993]-1" daemon [_thread_blocked, id=89, stack(0x00007fd81532c000,0x00007fd81542d000)]
  0x00007fd8604d9800 JavaThread "Service Thread" daemon [_thread_blocked, id=87, stack(0x00007fd815e48000,0x00007fd815f49000)]
  0x00007fd8604d6800 JavaThread "C1 CompilerThread14" daemon [_thread_blocked, id=86, stack(0x00007fd815f49000,0x00007fd81604a000)]
  0x00007fd8604d4800 JavaThread "C1 CompilerThread13" daemon [_thread_blocked, id=85, stack(0x00007fd81604a000,0x00007fd81614b000)]
  0x00007fd8604d2800 JavaThread "C1 CompilerThread12" daemon [_thread_blocked, id=84, stack(0x00007fd81614b000,0x00007fd81624c000)]
  0x00007fd8604d0800 JavaThread "C1 CompilerThread11" daemon [_thread_blocked, id=83, stack(0x00007fd81624c000,0x00007fd81634d000)]
  0x00007fd8604ce800 JavaThread "C1 CompilerThread10" daemon [_thread_blocked, id=82, stack(0x00007fd81634d000,0x00007fd81644e000)]
  0x00007fd8604cc800 JavaThread "C2 CompilerThread9" daemon [_thread_blocked, id=81, stack(0x00007fd81644e000,0x00007fd81654f000)]
  0x00007fd8604ca000 JavaThread "C2 CompilerThread8" daemon [_thread_blocked, id=80, stack(0x00007fd81654f000,0x00007fd816650000)]
  0x00007fd8604c8000 JavaThread "C2 CompilerThread7" daemon [_thread_blocked, id=79, stack(0x00007fd816650000,0x00007fd816751000)]
  0x00007fd8604c6000 JavaThread "C2 CompilerThread6" daemon [_thread_blocked, id=78, stack(0x00007fd816751000,0x00007fd816852000)]
  0x00007fd8604c4000 JavaThread "C2 CompilerThread5" daemon [_thread_blocked, id=77, stack(0x00007fd816852000,0x00007fd816953000)]
  0x00007fd8604ba000 JavaThread "C2 CompilerThread4" daemon [_thread_blocked, id=76, stack(0x00007fd816953000,0x00007fd816a54000)]
  0x00007fd8604b7800 JavaThread "C2 CompilerThread3" daemon [_thread_blocked, id=75, stack(0x00007fd816a54000,0x00007fd816b55000)]
  0x00007fd8604b5800 JavaThread "C2 CompilerThread2" daemon [_thread_blocked, id=74, stack(0x00007fd816b55000,0x00007fd816c56000)]
  0x00007fd8604b3800 JavaThread "C2 CompilerThread1" daemon [_thread_blocked, id=73, stack(0x00007fd816c56000,0x00007fd816d57000)]
  0x00007fd8604b1000 JavaThread "C2 CompilerThread0" daemon [_thread_blocked, id=72, stack(0x00007fd816d57000,0x00007fd816e58000)]
  0x00007fd8604af800 JavaThread "Signal Dispatcher" daemon [_thread_blocked, id=71, stack(0x00007fd816e58000,0x00007fd816f59000)]
  0x00007fd86047c800 JavaThread "Finalizer" daemon [_thread_blocked, id=70, stack(0x00007fd8170f1000,0x00007fd8171f2000)]
  0x00007fd860478000 JavaThread "Reference Handler" daemon [_thread_blocked, id=69, stack(0x00007fd8171f2000,0x00007fd8172f3000)]
```

**Comment by matiwinnetou** [comment]
@irbull 
Maybe error is in V8 itself and we should try Latest V8 (Latest NodeJS 8.x series) ?

Do you have other ideas?

**Comment by matiwinnetou** [comment]
Alternatively what if one tries to not to rely on classes that use:
```
i::Isolate* isolate = i::Isolate::Current();
```
and alternatively:
```
v8::Isolate::GetCurrent()
```

Perhaps it is possible to use other classes or even copy the code to J2V8 and reimplement those classes without relying on Isolate::Current. Explicitly passing Isolate as my PR: #308 and #310 indeed helped a bit, it crashes less frequently.

In the morning, I am gonna try again in PROD - this time we gonna use old legacy system based on Debian Jessie and redirect the traffic, this system is running on VMWare instead of docker images on Ubuntu 16.04, which is a system shared with lots of others apps. There is a possibility that this sharing can cause issues (although low probability).

**Comment by matiwinnetou** [comment]
I am also gonna try tomorrow NodeJS 7.10.1 build which has a bit newer version of V8. This is not as easy as I though. Build works on macos but not on linux.

**Comment by matiwinnetou** [comment]
@irbull what modifications did you have to do to NodeJS 7.4.x in order for it to be built successfully?

I noticed that nodejs gzip file is a special one downloaded from eclipsesource download page.

**Comment by matiwinnetou** [comment]
@irbull 

I am leaning towards bugs in V8 with concurrency, although we could also try V8 folks to do some concurrency review of J2V8 (if any of them would really wish to do this). This product does not seem to be battle tested (V8 in terms of concurrency / isolates).

We were brain storming today about this and we think that:
1. It would make sense to create a PR to V8 adding extra constructor to String::String::Value that takes isolate as a parameter, see: https://github.com/matiwinnetou/J2V8/commit/9b34493c71e5576ee20b171acdc68b71a4b6ae0e
This would also mean that implementation should be provided for this constructor in V8 itself. Even if it is accepted and merged to V8 it will take a long time for NodeJS to receive it but eventually it it could happen.

2. Do you have an idea how to implement some validation method, since we have a pool of v8 workers that one of them crashes every 6 hours now (due to this issue), perhaps if we were able to correctly and better implement validate method from the pool, on either creation, return to the pool of getting from the pool we could prevent the crash, mark the object as invalid and force recreation of it.
Do you think we could detect that those isolates desynchronized between our runtime and V8 or in fact when Isolate::Current() or Isolate::GetCurrent() returns NULL, then we would kill that instance of V8?

Our current implementation of valid method is very simple since we don't really have hooks how to detect that V8 instance is in desync and incorrect state:

```
    public boolean isValid() {
        return !isReleased();
    }

    private boolean isReleased() {
        if (polyfillObject instanceof V8Value) {
            V8Value polyfillObjectV8Value = (V8Value) this.polyfillObject;
            if (polyfillObjectV8Value.isReleased()) {
                return false;
            }
        }
        if (serverPageObject instanceof V8Value) {
            V8Value serverPageObjectV8Value = (V8Value) this.serverPageObject;
            if (serverPageObjectV8Value.isReleased()) {
                return false;
            }
        }

        return v8.isReleased() || compileTimeGlobals.isReleased();
    }
```

**Comment by irbull** [comment]
Going back through some of the questions here:

1. The version of Node that is downloaded from EclipseSource has had no changes to it. It's simply a built version of Node. It takes about 1 hour to build Node.js and the Travis-ci often times out when we build it as part of our build. Instead, we have a pre-built version which we download.

2. Yes, I think it would make sense to update V8 to include a fix for `v8::String::Value::Value(v8::Local<v8::Value>` and ideally anywhere else we use an API that tries to get the current Isolate. I know there are subtle memory issues when working with singletons in Java, and the same may be true for C/C++. So anything that relies on GetCurrent should be removed.

3. I don't know how to validate that the isolate is not desynchronized. We could add a static method to V8 called `getCurrent` that uses the native GetCurrent implementation. Then you could check that it doesn't return null? If we could do this, then the question is, could we reconnect things automatically?

**Comment by matiwinnetou** [comment]
ad 3. Have a look what I am cooking:
https://github.com/matiwinnetou/J2V8/tree/is-valid-isolate, I will try to test this tomorrow, do you see this working even?

**Comment by matiwinnetou** [comment]
ad 2 I am already working on a v8 contribution but the process to contribute to v8 is far from simple.

**Comment by matiwinnetou** [comment]
ad 1. My blind guess is that maybe things are improved in 7.6.0 (it builds out of the box) which uses new v8, but the guess is pretty blind and wishful thinking.

Why do we need extra patches for 7.4.0, this debug info supressing?

**Comment by matiwinnetou** [comment]
One thing for sure, the cases where we now explicitly pass Isolate do no longer produce exceptions in PROD, ie. the pull requests you merged in.

**Comment by matiwinnetou** [comment]
Last but not least, your improvements inside of branch: isolate_enter do not produce any side effects, but the do not fix the String::String::Value exception.

**Comment by irbull** [comment]
What are the extra patches you are referring to in "Why do we need extra patches for 7.4.0, this debug info supressing?"?

Regarding the isolate_enter branch, I tend to agree. It doesn't help, but it doesn't hurt either. I can get that in. 

I just created a test case that created 4M threads and acquired and released locks on different V8 instances to see if I could reproduce this. Nothing yet 😢.

I was looking back through your stack traces, and I wonder, are you releasing the locker during a callback? If you have any V8 code on your stack, you can't really release the lock (although I think it would fail with the error cannot exit an non-entered isolate (because when it tries to return back and deconstruct the Scope, it cannot `exit()`).

**Comment by matiwinnetou** [comment]
I absolutely believe you in terms of reproduction, I can only reproduce this in PROD and never on test environment, which makes it hard to play with it on QA environment. Isolate Enter bug #306  I can really easily reproduce but this one is now my main focus now but later after this one is nailed or mitigated. The instance now blows up usually after 8 hours but some run like 20 hours without blow up. SiteOps accepts this a bit more now but it is still not a green integration (more like yellow).

I will have a look into your comments, in fact I don't think it will be a problem to share a couple of classes we have with you. Only if you would have time look into this (integration of J2V8). On that note I was wondering if somebody from Google V8 team could review our code in terms of those concurrency issues.

I am referring to diff file in the project node.patches file: 7.4.0.diff

Do you think the idea which I have with isValidIsolate() method something that could work? If yes, then I 
really don't care that much that one V8 spontainously from the pool needs to be recreated every 8 hours (as it went invalid).

I am gonna look tomorrow into your suggestions and as I said provided you would want to review it, sharing some code shouldn't be a big deal. (I hope). Gonna check it tomorrow.

If it does result in success, not only this is an awesome Talk or block article but there would a first real usage of react SSR with J2V8. Lots of people think that NodeJS is a must where as this project / real serious site use case we have, could prove them otherwise.

**Comment by matiwinnetou** [comment]
@irbull I found locally some issues with isolate_enter_branch before I really understand what is going on please do not merge this to master.

**Comment by matiwinnetou** [comment]
@irbull Another day another thoughts and investigations. Crashing a JVM is a serious thing but failing a request for the user and keeping the JVM healthy is less serious. Would you accept a PR from us (or we have to run a branch), in which we try and catch exception that can happen during invocation of function, etc (request scoped invocation in terms of web request) and if the exception happens, catch this in C++ and then rethrow to java layer. Java layer could then should an error page and ask user to refresh the page. While catching the exception the V8 instance could be somehow invalidated or killed which client code could control.

This is of course not the solution to the problem but it reduces the criticality of a failure a bit. Since you are not observing those failures I don't know how comfortable would you feel in trusting us that they might indeed be race conditions failures in J2V8 and as I said recently even more likely in V8 thread safety code.

In the meantime we could still continue to work on finding if there is a way to improve this better but stopping JVM crash already would improve the stability for us to the next level.

My preliminary first code about checking the health of isolates didn't work but I will spend more time on this, my day was very packed today with other items. I will report this as well. So far I saw very weird stuff. v8::Isolate::GetCurrent() was NULL when everything was fine... and it should not be NULL at all, etc.

**Comment by irbull** [comment]
There is a lot going on here, so hopefully I haven't missed anything. Please re-ask if you're waiting for an answer.

Regarding the node build, it's just the output of https://github.com/eclipsesource/static-node-build. I don't think any patches are applied (unless the node build does that). I just run the `./buildAll` and take the output as a tarball and feed that into the J2V8 build. This just keeps us from building node each time.

Regarding throwing an exception if things are off, tis seems ok. But it's hard to know if V8 can self correct and we would be throwing an exception for no reason. 

I've been thinking more about the root cause, and I wonder (as mentioned above) if it's related to re-entrant code. Consider the following:
V8_Instance1.callFunction();
  -> This calls a Java function and clearly V8_Instance1 is the _currentIsolate_.
  -> During this call several new Isolates are created. Now _what is the currentIsolate_?
  -> Some of those new isolates are removed. Now _what is the currentIsolate_?
  -> callFunction() returns. As part of the return, SetValue is called. What is currentIsolate?

In the finally step (the return of callFunction), V8_Instance1 needs to be the currentIsolate, but how do we ensure this? We create the scope when callFunction is invoked, and never again (because it's still on the stack). But I could imagine that V8 will not be able to track that. It's easy for us to know, and we can also pass the current `Isolate`, but only if V8 has the appropriate methods.

Finally, regarding your node about node.js. Node isn't thread based, it's process based. If you want to scale node, you add more processes. So this not battle tested from a node perspective. Webworkers in Chrome use threads, but likely not to the extent you are.

**Comment by matiwinnetou** [comment]
Thanks @irbull, yes maybe went I said NodeJS I really meant V8, I know the single threading model of NodeJS (just silly mistakes).

Exactly, we can only pass current isolate if V8 has methods and constructors, which as we already know it does not always have. I will probably open PR to V8 code anyway but now I am trying to solve immediately the problem and if they even approve this PR and merge this, it will take long time for J2V8 to receive this version of V8.

I am not sure I follow 100% what you say but basically the only other option for me to brand J2V8 as success level for us on the JVM is only two options, the other ones are already way bigger hacks (like catch and re-throwing exception - obvious workaround because user sees error page, evicting V8 from the pool based on time strategy of number of usages of pool entry - obvious workaround, how often should we evict?):

Options for some level of success for future:
1. V8 java instance and isValid() or isHealthy() method, as I shared the implementation already above in previous comments, so far it doesn't work - if this would work we would be able to recreate V8 instance on borrowing from the pool before the request crashes JVM or results in 500 error code.

2. Usage of J2V8 for development mode only (very fast - 500 ms to compile the whole server side bundle and show the page) and usage of Nashorn in production (takes at least 5 secs to preheat one of our server side bundle Nashorn instance). Do you have here any experiences of using Nashorn for web development for a relatively serious site in terms of traffic?

3. Usage dedicated 1 Thread per JVM instance. That way V8 would ever be only accessed by one thread. Still not so easy not only because we will needs lots of small JVM instances but also because java servlets are traditionally blocking...

I don't have other ideas now to make success of this further (React SSR with J2V8) and use this a a blueprint to other pages and not abandoning it - we use this on isolated small page but we wanna rewrite other pages without rewriting the whole webapp to NodeJS. Are there any hints you could give us for work item (1), isHealthy() or isValid(long ptr_Runtime) method on V8, how could we implement if V8 is on java tier is healthy... and including underlying (or especially) V8 C++ instance.

**Comment by matiwinnetou** [comment]
Oh and option 4, magically hoping in a wishful thinking manner that newer version of NodeJS have such a V8 that fixes Isolate issues... (probability success: low)

**Comment by matiwinnetou** [comment]
@irbull Thank you so much for your ongoing help as well. I don't take this for granted and would be also lovely to donate to J2V8.

**Comment by matiwinnetou** [comment]
I completely forgot about one more option, patching V8 inside of J2V8 codebase and adding extra constructors with implementation:
Work item which I started here with experiments:
https://github.com/matiwinnetou/J2V8/commit/9b34493c71e5576ee20b171acdc68b71a4b6ae0e, commit is not 100% full because there is also change in api.cc -> adding two new constructors that take isolate. This would have to be added to:
![image](https://user-images.githubusercontent.com/335933/28693944-7db4ee10-7327-11e7-96ee-9c913016259b.png)

**Comment by matiwinnetou** [comment]
We went live today with fixes in V8. Adding Isolate* to two constructors in V8.h and implementation in api.cc.

https://github.com/matiwinnetou/J2V8/commit/30fba14f5c2291fafa2eed5bd4042f9446cafb49

Already tomorrow we will know if the issue is fixed. (it took before about 6-8 hours in PROD to see a crash caused by Isolate desync.)

**Comment by matiwinnetou** [comment]
I just checked, all webapps ran without a String::String::Value crash for the whole night - in total 16 hours. We never achieved result like this - usually crashed after 8 hours. I will prepare a patch and add to this project in node.patches folder, then open a PR. If PR is not accepted then we will be forced to ran a branch of the project and rebase regularly from master. I know no-one wants to have patches...

I will also prepare a PR to V8 project this weekend, explicitly passing Isolate to constructors of String::Value and UtfValue...

**Comment by matiwinnetou** [comment]
@irbull Do you anticipate that your latest changes in master could in fact fix this issue or the issue is in V8 and my patches to V8 where Isolate is passed explicitly are still the best solution to the problem?

**Comment by matiwinnetou** [comment]
https://docs.google.com/document/d/1g8JFi8T_oAE_7uAri7Njtig7fKaPDfotU6huOa1alds/edit

**Comment by matiwinnetou** [comment]
Onoing or planned changes:
```
Force explicit Isolate* parameter on all external APIs: bug
The idea is to get rid of internal TLS dependencies. Instead, the embedder has to always specify which Isolate to use. 
```

**Comment by irbull** [comment]
I have reason to believe that the fix for #321 may actually finally fix this issue. Let me explain.

The root problem is that `Isolate::GetCurrent` does not in fact point to the _current_ isolate. We track the _current_ isolate in all J2V8 JNI calls. `Isolate::GetCurrent` is managed by `Isolate::Enter()` and `Isolate::Exit()`, which are set via a new `Isolate::Scope` each time we enter a JNI call, so the question I've been struggling with is _how can `GetCurrent` be wrong?

J2V8 is meticulous about setting up a scope (and tearing it down) on each JNI call, which should enforce that the current isolate is always correct. However, there was one case were we did not do this. During Isolate creation, we created a Scope dynamically on the heap and tore it down during the release. This meant that whenever an isolate was created or released the pattern changed. If this was done during a callback or some other JS execution, then we could end up with the wrong current isolate. 

Looking at all the stack traces you provided, it always appeared to happen after a JS function call or script execution, fuelling my theory.

With the changes to fix #321, we no longer create those Scopes on the heap, and like all other Scopes they are stack based. This means that J2V8 will now *always* enter an isolate on each JNI call, and exit from that isolate when the call returns. I am hopeful this will finally fix the problem.

**Comment by irbull** [comment]
I do think that explicitly stating our Isolate is the best approach, and if the V8 team is moving towards that for all their APIs then we should use that. I don't mind patching V8/Node during our build, but I find that those types of patches become a maintenance headache as we move forward with the project. If we could fix this without patching, that would be ideal. 

Even if this does in fact fix the issue, please continue to work with the V8 team to get your patch in. That's the right solution in the long run, IMHO.

**Comment by matiwinnetou** [comment]
Thanks, now that I have a fixed patched version proved to work, after basic testing in QA environment I will deploy latest master to PROD and report to you if it worked.

**Comment by matiwinnetou** [comment]
I am getting there: https://chromium-review.googlesource.com/c/593309/ but this gerrit review is still wrong, now I need to figure out how to fix silly mistakes.

**Comment by irbull** [comment]
i'll take a look at your V8 patch tomorrow. I've worked with Gerrit a bunch, and even contributed a small change to the gerrit project itself (which of course uses Gerrit).

**Comment by matiwinnetou** [comment]
Sure, I also worked with Gerrit in the past but I mostly forgot all stuff :)

**Comment by matiwinnetou** [comment]
Looks like https://github.com/eclipsesource/J2V8/commit/6c71f12a6c95eb564967e731c764ffd6bb9a095a fixed it. Tomorrow morning, I will know for sure but it already looks very promising. No single crash yet.

If all is fine, I will close this issue tomorrow. When this is closed I would suggest to release a new version of J2V8.

**Comment by irbull** [comment]
🤞 🤞 🤞 🤞 🤞 🤞 🤞 
Let's hope.

**Comment by matiwinnetou** [comment]
9 hours without any crash in neither data centers, looks nailed. Gonna close this in the morning (CEST timezone). 

Gerrit review completed, I applied all comments from the project maintainer: Adam Klein:

https://chromium-review.googlesource.com/c/593309

**Comment by irbull** [comment]
Once this makes it way into a V8 version we can consume, we should make sure we use this API. I opened #325 to track this.

**Comment by matiwinnetou** [comment]
16 hours no crash, fixed in master, closing.
Awesome work @irbull.

**Comment by irbull** [comment]
@matiwinnetou Thank-you for pushing me on this, all the investigation you did and all testing (and all your patience). Thanks for working with the V8 team on this as well. Great work!
